File: test/hardware/unit/test_branch_unit.py
import cocotb
from cocotb.triggers import Timer
import random
from pathlib import Path
from test.driver import run_hardware_test

@cocotb.test()
async def branch_unit_test(dut):
    """Test Branch Unit."""

    # Helper to check output
    async def check(funct3, a, b, expected, name):
        dut.function_3.value = funct3
        dut.operand_a.value = a
        dut.operand_b.value = b
        await Timer(1, units="ns")
        
        got = int(dut.branch_condition_met.value)
        assert got == expected, f"{name} Failed: A={hex(a)}, B={hex(b)}, Expected={expected}, Got={got}"

    # Test Cases
    # BEQ (000)
    await check(0b000, 10, 10, 1, "BEQ Equal")
    await check(0b000, 10, 20, 0, "BEQ Not Equal")

    # BNE (001)
    await check(0b001, 10, 20, 1, "BNE Not Equal")
    await check(0b001, 10, 10, 0, "BNE Equal")

    # BLT (100) - Signed
    await check(0b100, -10 & 0xFFFFFFFF, 10, 1, "BLT Less")
    await check(0b100, 10, -10 & 0xFFFFFFFF, 0, "BLT Greater")
    await check(0b100, 10, 10, 0, "BLT Equal")

    # BGE (101) - Signed
    await check(0b101, 10, -10 & 0xFFFFFFFF, 1, "BGE Greater")
    await check(0b101, 10, 10, 1, "BGE Equal")
    await check(0b101, -10 & 0xFFFFFFFF, 10, 0, "BGE Less")

    # BLTU (110) - Unsigned
    await check(0b110, 10, 20, 1, "BLTU Less")
    await check(0b110, 0xFFFFFFFF, 10, 0, "BLTU Greater (Unsigned)") # -1 is huge in unsigned

    # BGEU (111) - Unsigned
    await check(0b111, 0xFFFFFFFF, 10, 1, "BGEU Greater (Unsigned)")
    await check(0b111, 10, 10, 1, "BGEU Equal")
    await check(0b111, 10, 20, 0, "BGEU Less")

    dut._log.info("Branch Unit Test Passed!")

def test_branch_unit():
    run_hardware_test(
        module_name=Path(__file__).stem,
        toplevel="branch_unit",
        verilog_sources=["core/backend/branch_unit.v"]
    )

File: test/hardware/unit/test_branch_unit.cpp
#include "tb_base.h"
#include "Vbranch_unit.h"
#include <iostream>

// Branch function codes
const uint8_t FUNCT3_BEQ  = 0b000;
const uint8_t FUNCT3_BNE  = 0b001;
const uint8_t FUNCT3_BLT  = 0b100;
const uint8_t FUNCT3_BGE  = 0b101;
const uint8_t FUNCT3_BLTU = 0b110;
const uint8_t FUNCT3_BGEU = 0b111;

class BranchUnitTestbench : public TestbenchBase<Vbranch_unit> {
public:
    BranchUnitTestbench() : TestbenchBase<Vbranch_unit>(true, "branch_unit_trace.vcd") {
        TB_LOG("Branch Unit Testbench initialized");
    }
    
    void test_branch(uint8_t funct3, uint32_t a, uint32_t b, bool expected, const char* name) {
        dut->function_3 = funct3;
        dut->operand_a = a;
        dut->operand_b = b;
        eval();
        
        bool actual = dut->branch_condition_met;
        if (actual != expected) {
            std::cerr << "FAIL: " << name << "(0x" << std::hex << a << ", 0x" << b 
                      << ") = " << actual << ", expected " << expected << std::dec << std::endl;
            throw std::runtime_error("Branch test failed");
        }
    }
    
    void run_all_tests() {
        TB_LOG("Running Branch Unit tests...");
        
        // BEQ tests
        test_branch(FUNCT3_BEQ, 10, 10, true, "BEQ");
        test_branch(FUNCT3_BEQ, 10, 20, false, "BEQ");
        
        // BNE tests
        test_branch(FUNCT3_BNE, 10, 20, true, "BNE");
        test_branch(FUNCT3_BNE, 10, 10, false, "BNE");
        
        // BLT tests (signed)
        test_branch(FUNCT3_BLT, 5, 10, true, "BLT");
        test_branch(FUNCT3_BLT, 10, 5, false, "BLT");
        test_branch(FUNCT3_BLT, 0xFFFFFFFF, 1, true, "BLT");  // -1 < 1
        
        // BGE tests (signed)
        test_branch(FUNCT3_BGE, 10, 5, true, "BGE");
        test_branch(FUNCT3_BGE, 5, 10, false, "BGE");
        test_branch(FUNCT3_BGE, 10, 10, true, "BGE");
        
        // BLTU tests (unsigned)
        test_branch(FUNCT3_BLTU, 5, 10, true, "BLTU");
        test_branch(FUNCT3_BLTU, 0xFFFFFFFF, 1, false, "BLTU");  // 0xFFFFFFFF > 1
        
        // BGEU tests (unsigned)
        test_branch(FUNCT3_BGEU, 10, 5, true, "BGEU");
        test_branch(FUNCT3_BGEU, 0xFFFFFFFF, 1, true, "BGEU");
        
        TB_LOG("All Branch Unit tests PASSED");
    }
};

int main(int argc, char** argv) {
    Verilated::commandArgs(argc, argv);
    
    try {
        BranchUnitTestbench tb;
        tb.run_all_tests();
        
        TB_LOG("==================================" );
        TB_LOG("All Branch Unit tests PASSED!");
        TB_LOG("==================================");
        return 0;
    } catch (const std::exception& e) {
        TB_ERROR(e.what());
        return 1;
    }
}
